### PR TITLE
feat(log_store): introduce the CollectionTask

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)

#4385
## What's changed and what's your intention?

Introduce the `CollectionTask`, it performs two main operations:
- Persists the WAL index to the specified `path` at every `dump_index_interval`.
- Updates the latest index ID for each WAL provider at every `checkpoint_interval`.

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
	- Introduced enhanced configuration options for indexing, including new parameters for index creation and intervals.
	- Added a new `GlobalIndexCollector` for better index management in the data node.
	- Expanded public API with new exports like `default_index_file` and `GlobalIndexCollector`.

- **Bug Fixes**
	- Improved error handling with more specific error messages for different operations.

- **Chores**
	- Updated dependency configurations to support new data handling and processing features.

### CODE REVIEW

**Comment by WenyXu** [comment]
@coderabbitai summary

File: src/log-store/src/kafka/index/collector.rs [hunk 1]
```diff
@@ -58,23 +52,175 @@ pub trait IndexCollector: Send + Sync {
...
+                    provider.topic
+                )
+            }
+        }
+    }
+
+    /// The background task performs two main operations:
+    /// - Persists the WAL index to the specified `path` at every `dump_index_interval`.
+    /// - Updates the latest index ID for each WAL provider at every `checkpoint_interval`.
+    fn run(&mut self) {
```
Comment (fengjiachun) [inline src/log-store/src/kafka/index/collector.rs:123]: `&mut` is not necessary

File: src/log-store/src/kafka/index/collector.rs [hunk 2]
```diff
@@ -58,23 +52,175 @@ pub trait IndexCollector: Send + Sync {
...
+    path: String,
+    running: Arc<AtomicBool>,
+}
+
+impl CollectionTask {
+    async fn dump_index(
+        providers: &Arc<TokioMutex<HashMap<Arc<KafkaProvider>, Sender<WorkerRequest>>>>,
+        operator: &object_store::ObjectStore,
+        path: &str,
+    ) -> Result<()> {
```
Comment (fengjiachun) [inline src/log-store/src/kafka/index/collector.rs:76]: why not `dump_index(&self)`?

**Comment by WenyXu** [comment]
@fengjiachun @v0y4g3r PTAL

File: src/common/wal/src/config/kafka/datanode.rs [hunk 3]
```diff
@@ -40,6 +40,11 @@ pub struct DatanodeKafkaConfig {
     /// The kafka topic config.
     #[serde(flatten)]
     pub kafka_topic: KafkaTopicConfig,
+    pub create_index: bool,
+    #[serde(with = "humantime_serde")]
+    pub dump_index_interval: Duration,
+    #[serde(with = "humantime_serde")]
+    pub index_checkpoint_interval: Duration,
```
Comment (fengjiachun) [inline src/common/wal/src/config/kafka/datanode.rs:47]: Maybe we can merge dump_index and index_checkpoint?

Comment (WenyXu) [reply src/common/wal/src/config/kafka/datanode.rs]: Resolved in [73ac3b1](https://github.com/GreptimeTeam/greptimedb/pull/4530/commits/73ac3b14788ee06b87c84bb5800e902687120a30)

**Comment by fengjiachun** [comment]
@WenyXu CI failed

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Pull request merged**